### ISSUE TITLE
AST-based normalization / denormalization

### ISSUE THREAD

**Issue by smkhalsa** [zino-app/graphql-flutter#471]
With the addition of AST-based operations in #429, I'd like to propose that we implement AST-aware cache normalization / denormalization.

This will allow us to only read the requested fields from the cache, which will allow us to get rid of all the [dereferencing](https://github.com/zino-app/graphql-flutter/blob/master/packages/graphql/lib/src/cache/lazy_cache_map.dart) code.

It will also make it trivial to implement readQuery / readFragment & writeQuery / writeFragment as requested in #176.

I just published [normalize](https://pub.dev/packages/normalize), a GraphQL AST-based normalization library as a first step. I've [tested](https://github.com/smkhalsa/normalize/tree/master/test) it for all the common use cases, and it currently supports:

1. Fragments (definitions & inline)
2. Variables
3. Unions & Interface types
4. Type Aliases
5. Apollo's new TypePolicies (#458)

Per @klavs's [suggestion](https://github.com/zino-app/graphql-flutter/issues/400#issuecomment-539147245), it also addresses #400 as it serializes query fields with their variables (like Apollo does, with optional field-level config via FieldPolicies).

Please let me know your thoughts.

**Comment by micimize** [comment]
> which will allow us to get rid of all the dereferencing code.

Oh man, finally looking at this and it's super exciting. I can't wait to delete `LazyCacheMap` now lol

**Comment by SwannLV** [comment]
Can't wait to have aliases supported in my cache. Could I use it now?

**Comment by SwannLV** [comment]
Would it get the cache by looking at if all fields are in it? 
I miss that a lot in my project.
Thank you very much 🙏

**Comment by micimize** [comment]
As mentioned in #629, this will also solve the broken cache behavior of 
```graphql
query queryA {
   abc {
      fieldA, fieldB
   }
}

# resolves the partial from `queryA` which is entirely incorrect
query queryB {
   abc {
      fieldC, fieldD
   }
}
```

**Comment by micimize** [comment]
closed with `4.0.0-beta.1` 🙌  I still need to migrate us to normalize `^0.4.0` though 😅
